demos/sslecho/main.c

```diff
@@ -0,0 +1,340 @@
+ */
+static volatile bool    server_running = true;
+
+int create_socket(bool isServer)
+{
+    int s;
+    int optval = 1;
+    struct sockaddr_in addr;
+
+    s = socket(AF_INET, SOCK_STREAM, 0);
+    if (s < 0) {
+        perror("Unable to create socket");
+        exit(EXIT_FAILURE);
+    }
+
+    if (isServer) {
+        addr.sin_family = AF_INET;
+        addr.sin_port = htons(server_port);
+        addr.sin_addr.s_addr = INADDR_ANY;
+
+        /* Reuse the address; good for quick restarts */
+        if (setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval))
+                < 0) {
+            perror("setsockopt(SO_REUSEADDR) failed");
+            exit(EXIT_FAILURE);
+        }
+
+        if (bind(s, (struct sockaddr*) &addr, sizeof(addr)) < 0) {
+            perror("Unable to bind");
+            exit(EXIT_FAILURE);
+        }
+
+        if (listen(s, 1) < 0) {
+            perror("Unable to listen");
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    return s;
+}
+
+SSL_CTX* create_context(bool isServer)
+{
+    const SSL_METHOD *method;
+    SSL_CTX *ctx;
+
+    if (isServer)
+        method = TLS_server_method();
+    else
+        method = TLS_client_method();
+
+    ctx = SSL_CTX_new(method);
+    if (ctx == NULL) {
+        perror("Unable to create SSL context");
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+
+    return ctx;
+}
+
+void configure_server_context(SSL_CTX *ctx)
+{
+    /* Set the key and cert */
+    if (SSL_CTX_use_certificate_chain_file(ctx, "cert.pem") <= 0) {
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+
+    if (SSL_CTX_use_PrivateKey_file(ctx, "key.pem", SSL_FILETYPE_PEM) <= 0) {
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+}
+
+void configure_client_context(SSL_CTX *ctx)
+{
+    /*
+     * Configure the client to abort the handshake if certificate verification
+     * fails
+     */
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+    /*
+     * In a real application you would probably just use the default system certificate trust store and call:
+     *     SSL_CTX_set_default_verify_paths(ctx);
+     * In this demo though we are using a self-signed certificate, so the client must trust it directly.
+     */
+    if (!SSL_CTX_load_verify_locations(ctx, "cert.pem", NULL)) {
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+}
+
+void usage()
+{
+    printf("Usage: sslecho s\n");
+    printf("       --or--\n");
+    printf("       sslecho c ip\n");
+    printf("       c=client, s=server, ip=dotted ip of server\n");
+    exit(1);
+}
+
+int main(int argc, char **argv)
+{
+    bool isServer;
+    int result;
+
+    SSL_CTX *ssl_ctx = NULL;
+    SSL *ssl = NULL;
+
+    int server_skt = -1;
+    int client_skt = -1;
+
+    char *txbuf = (char*) OPENSSL_malloc(128);
+    size_t txcap = 128;
+    int txlen;
+
+    char *rxbuf = (char*) OPENSSL_malloc(128);
```

Comment:
`txbuf` and `rxbuf` ought to be checked for NULL sometime later or made fixed sized arrays:
```C
char txbuf[128];
size_t txcap = sizeof(txbuf);
int txlen;

char rxbuf[128];
size_t rx-cap = sizeof(rxbuf);
int rxlen;
```

Reply:
This use of getline with fixed size array causes incompatible pointer type warning.

```
cc -O0 -g3 -W -Wall -I../../include -L../../ -o sslecho main.c -lssl -lcrypto
main.c: In function ‘main’:
main.c:288:33: warning: passing argument 1 of ‘getline’ from incompatible pointer type [-Wincompatible-pointer-types]
  288 |                 txlen = getline(&txbuf, &txcap, stdin);
      |                                 ^~~~~~
      |                                 |
      |                                 char (*)[128]
In file included from main.c:10:
/usr/include/stdio.h:616:45: note: expected ‘char ** restrict’ but argument is of type ‘char (*)[128]’
  616 | extern __ssize_t getline (char **__restrict __lineptr,
      |                           ~~~~~~~~~~~~~~~~~~^~~~~~~~~

```

Reply:
this change broke the demo. Geting it back to work with 

```
diff --git a/demos/sslecho/main.c b/demos/sslecho/main.c
index c6f087c56f..93600e2c84 100644
--- a/demos/sslecho/main.c
+++ b/demos/sslecho/main.c
@@ -285,7 +285,8 @@ int main(int argc, char **argv)
             /* Loop to send input from keyboard */
             while (true) {
                 /* Get a line of input */
-                txlen = getline(&txbuf, &txcap, stdin);
+                char * txbufp = txbuf;
+                txlen = getline(&txbufp, &txcap, stdin);
                 /* Exit loop if just a carriage return */
                 if (txbuf[0] == '\n') {diff --git a/demos/sslecho/main.c b/demos/sslecho/main.c
index c6f087c56f..93600e2c84 100644
--- a/demos/sslecho/main.c
+++ b/demos/sslecho/main.c
@@ -285,7 +285,8 @@ int main(int argc, char **argv)
             /* Loop to send input from keyboard */
             while (true) {
                 /* Get a line of input */
-                txlen = getline(&txbuf, &txcap, stdin);
+                char * txbufp = txbuf;
+                txlen = getline(&txbufp, &txcap, stdin);
                 /* Exit loop if just a carriage return */
                 if (txbuf[0] == '\n') {
                     break;

                     break;

```

make

# starting server
LD_LIBRARY_PATH=../..:$LD_LIBRARY_PATH ./sslecho s&

# starting client
LD_LIBRARY_PATH=../..:$LD_LIBRARY_PATH ./sslecho c localhost
```

sslecho : Simple Echo Client/Server (OpenSSL 3.0.1-dev) : Apr 24 2022 : 19:13:47

We are the client

TCP connection to server successful
Client TCP connection accepted
SSL connection to server successful

Client SSL connection accepted

test
Received: test
Received: test

```